### ISSUE TITLE
Clean up methods in LEDMode class

### ISSUE THREAD

**Issue by cdisselkoen** [keyboardio/Kaleidoscope-LEDControl#9]
Currently LEDMode has three separate methods called begin(), setup(), and init(), and it's not clear what the distinction is between them, or how new LED modes should choose which one (or more) of them to override.

**Comment by cdisselkoen** [comment]
Separately, LEDControl has begin(), setup(), and init_mode().  These do not correspond, as (for instance) LEDControl.begin() calls LEDMode.setup().  We should probably make these correspond more nicely.

**Comment by cdisselkoen** [comment]
Here are my current thoughts.  In addition to renaming methods, I thought I'd formalize and document the expected 'contract' between `LEDControl` and `LEDMode` a little better.

In the process, the addition of the `refresh()` methods is one possible solution for keyboardio/Kaleidoscope-LEDEffect-SolidColor#2 that doesn't require `SolidColor` to refresh every LED on every `update()`.  `refresh()` would have to be explicitly invoked by plugins like `ActiveModColor` when they wish to 'reset' a key, i.e. stop overriding its color.  If we decide on an alternate solution for keyboardio/Kaleidoscope-LEDEffect-SolidColor#2, the rest of this still applies as a suggestion, just without the `refresh()` methods.

I realize that some of this would be superseded by a solution to #10, but a lot of it is orthogonal.  I just wanted to put this suggestion out here, or perhaps as a building block to #10.  

(I would plan on putting all these comments into the code, except the notes about what the old method names were.)

```
class LEDMode : public KaleidoscopePlugin {
  friend class LEDControl_;
  protected:
    // These methods should only be called by LEDControl.

    // One-time setup, done at boot.
    virtual void setup(void) {}

    // This is called whenever we switch to this mode
    // LEDControl will blank all LEDs prior to calling this.
    // Also, the first update() will happen immediately after onActivate()
    //   finishes, so you don't need to duplicate anything from update().
    // [Note: old name 'init']
    virtual void onActivate(void) {}

    // This is called regularly while this is the active mode
    virtual void update(void) {}

    // This is called while we are the active mode, if another plugin has changed
    //   a key's color and wants us to reprocess it.
    // If we set all LEDs during every update() anyway, this can just remain a
    //   no-op, as it will be taken care of on next update().
    virtual void refreshAt(byte row, byte col) {}
    // This is the same, but for all keys
    virtual void refreshAll() {}

  public:
    // This makes us the active mode
    void activate(void) final;

    // This is the begin() inherited from KaleidoscopePlugin.
    // All LEDModes share an implementation, and it cannot be overridden.
    // Use setup() instead for one-time setup tasks.
    void begin(void) final;
};
```

For class `LEDControl_`:
- remove the `setup()` method, as it seems it was never even implemented, let alone called (am I correct?)
- add public `refreshAt(byte row, byte col)` and `refreshAll()`, which just call the corresponding method on the active mode

One of the larger changes to the various `LEDMode` plugins that this proposes, is that `begin()` be non-overrideable.  Many plugins currently override `begin()`.  However, (a) that defeats the purpose of `setup()` - and it seems better for the future that `LEDControl` have some control over the setup process; and (b) I dislike contracts like "you can override this function, but you *absolutely must* call the superclass function or equivalent".  This two-tier setup process (a `begin()` marked `final`, and a virtual `setup()` which is by default a no-op) is cleaner and harder to mess up for new LEDModes.

**Comment by algernon** [comment]
The `refresh()` method makes sense in the stacked case too. Static modes could then have a no-op update method, and `ActiveModColor` could call `LEDControl.refresh()` which would refresh only those keys, optimizing things further. In short, good idea!

Rest of the proposal looks great as well, thanks a lot for writing this all down!

**Comment by algernon** [comment]
I stared working on #10, while using your proposed LEDMode API above, and there's a few things I'm having issues with. It mostly has to do with blanking the LEDs on `onActivate`: lets say I have two LED modes, one does all its work at `onActivate`, the other has its own `update` method. When I turn on the second one, if `onActivate` blanks the LEDs, I will loose the first effect. Unless I call `onActivate` for all active modes whenever a new one activates, but that sounds counter-intuitive.

Instead, I'd just nuke `LEDMode.update`, and rely on `LEDMode.refreshAt` instead, and introduce `LEDMode.isStatic()`, which would return true if the mode is static (no need to refresh every cycle), false otherwise. If a mode is static, then `LEDControl.update()` will not call their `refreshAt()`, to save a few cycles.

On the other hand, as we are going through the LEDs only once, such optimisation may not be neccessary... unless... unless we skip the update altogether if we only have static modes. Or even better, we can have `LEDMode.needUpdate()`! Static modes could return `false`, others could decide what to return. In `LEDControl.update()`, we'd first loop through all active modes to see which one wants to update, and only loop through the LEDs if any of them do.

**Comment by algernon** [comment]
Okay, I spoke too soon. `LEDMode.needUpdate` is not going to work, it would make `ActiveModColor` far too complicated, for example.

**Comment by algernon** [comment]
So, having arrived to the conclusion that LED mode stacking may not be worth it, blanking the LEDs before `onActivate` makes much more sense. With a single mode active at a time, this is useful behaviour.
